File: server/datasets/python_source-code/Multiple-F1/type3_structural_34_merge_sort.py
def merge_blocks(left_arr, right_arr):
    out = []
    i = j = 0
    while i < len(left_arr) and j < len(right_arr):
        if left_arr[i] <= right_arr[j]:
            out.append(left_arr[i])
            i += 1
        else:
            out.append(right_arr[j])
            j += 1
    while i < len(left_arr):
        out.append(left_arr[i])
        i += 1
    while j < len(right_arr):
        out.append(right_arr[j])
        j += 1
    return out

def merge_sort(numbers_list):
    if len(numbers_list) <= 1:
        return numbers_list
    mid = len(numbers_list) // 2
    right_part = merge_sort(numbers_list[mid:])
    left_part = merge_sort(numbers_list[:mid])
    return merge_blocks(left_part, right_part)

File: server/datasets/java_source-code/Multiple-F1/type3_structural_34_merge_sort.java
public class Solution {
    private int[] mergeBlocks(int[] leftArr, int[] rightArr) {
        int[] out = new int[leftArr.length + rightArr.length];
        int i = 0, j = 0, k = 0;
        while (i < leftArr.length && j < rightArr.length) {
            if (leftArr[i] <= rightArr[j]) out[k++] = leftArr[i++];
            else out[k++] = rightArr[j++];
        }
        while (i < leftArr.length) out[k++] = leftArr[i++];
        while (j < rightArr.length) out[k++] = rightArr[j++];
        return out;
    }

    public int[] mergeSort(int[] arr) {
        if (arr == null || arr.length <= 1) return arr;
        int mid = arr.length / 2;
        int[] leftPart = new int[mid];
        int[] rightPart = new int[arr.length - mid];
        System.arraycopy(arr, 0, leftPart, 0, mid);
        System.arraycopy(arr, mid, rightPart, 0, arr.length - mid);
        
        rightPart = mergeSort(rightPart);
        leftPart = mergeSort(leftPart);
        return mergeBlocks(leftPart, rightPart);
    }
}
